Src/Support/Google.Apis.Auth/OAuth2/ImpersonatedCredential.cs

```diff
@@ -0,0 +1,176 @@
+            if (!(SourceCredential.UnderlyingCredential is UserCredential || SourceCredential.UnderlyingCredential is ServiceAccountCredential))
+            {
+                throw new InvalidOperationException("The underlying credential of source credential must be UserCredential or ServiceAccountCredential.");
+            }
+            SourceCredential = SourceCredential.CreateScoped(new string[] {GoogleAuthConsts.IamScope});
+            HttpClient.MessageHandler.Credential = SourceCredential.UnderlyingCredential as Http.IHttpExecuteInterceptor;
+            Options = initializer.Options;
+        }
+
+        /// <inheritdoc/>
+        IGoogleCredential IGoogleCredential.WithQuotaProject(string quotaProject) =>
+            new ImpersonatedCredential(new Initializer(this) { QuotaProject = quotaProject });
+
+        /// <inheritdoc/>
+        public override async Task<bool> RequestAccessTokenAsync(CancellationToken taskCancellationToken)
+        {
+            var bodyJson = new ImpersonationAccessTokenRequest
+            {
+                DelegateAccounts = Options.DelegateAccounts,
+                Scopes = Options.Scopes,
+                Lifetime = ((int)Options.Lifetime.TotalSeconds).ToString() + "s"
+            };
+            var body = NewtonsoftJsonSerializer.Instance.Serialize(bodyJson);
+            var content = new StringContent(body, Encoding.UTF8, "application/json");
+            
+            var response = await HttpClient.PostAsync(TokenServerUrl, content, taskCancellationToken).ConfigureAwait(false);
+            var newToken = await TokenResponse.FromHttpResponseAsync(response, Clock, Logger).ConfigureAwait(false);
+            Token = newToken;
+            return true;
+        }
+
+        /// <inheritdoc/>
+        public Task<OidcToken> GetOidcTokenAsync(OidcTokenOptions options, CancellationToken cancellationToken = default)
+        {
+            options.ThrowIfNull(nameof(options));
+            // If at some point some properties are added to OidcToken that depend on the token having been fetched
+            // then initialize the token here.
+            TokenRefreshManager tokenRefreshManager = null;
+            tokenRefreshManager = new TokenRefreshManager(ct => RefreshOidcTokenAsync(tokenRefreshManager, options, ct), Clock, Logger);
+            return Task.FromResult(new OidcToken(tokenRefreshManager));
+        }
+
+        private async Task<bool> RefreshOidcTokenAsync(TokenRefreshManager caller, OidcTokenOptions oidcTokenOptions, CancellationToken cancellationToken)
+        {
+            var bodyJson = new ImpersonationOIdCTokenRequest
+            {
+                DelegateAccounts = Options.DelegateAccounts,
+                Audience = oidcTokenOptions.TargetAudience,
+                IncludeEmail = true
+            };
+            var body = NewtonsoftJsonSerializer.Instance.Serialize(bodyJson);
```

Comment:
See my comment above for adding extension methods.

Reply:
same as above